internal/bigquery/batch/client.go

```diff
@@ -0,0 +1,185 @@
+		schema, csvOptions,
+	)
+}
+
+func newClient(client *bigquery.Client, dataSetID, tableID string, ignoreUnknownValues, autoDetect bool, sourceFormat bigquery.DataFormat, writeDisposition *bigquery.TableWriteDisposition, schema *bigquery.Schema, csvOptions *bigquery.CSVOptions) (*Client, error) {
+	if dataSetID == "" {
+		return nil, fmt.Errorf("BQ batch client: validate dataSetID: %w: missing", internal.InvalidParamErr)
+	}
+	if tableID == "" {
+		return nil, fmt.Errorf("BQ batch client: validate tableID: %w: missing", internal.InvalidParamErr)
+	}
+
+	// If autoDetect is passed while the format is not JSON or CSV, error as this is only supported for csv.
+	if autoDetect && (sourceFormat != bigquery.JSON && sourceFormat != bigquery.CSV) {
+		return nil, autoDetectAndSchemaNotSupported
+	}
+
+	// If autoDetect is passed and both the schema is passed error as these are mutually exclusive.
+	if autoDetect && schema != nil {
+		return nil, autoDetectSchemaNotSupported
+	}
+
+	if !autoDetect && schema == nil {
+		return nil, schemaRequired
+	}
+
+	// csvOptions is only allowed if the sourceFormat is CSV.
+	if csvOptions != nil && sourceFormat != bigquery.CSV {
+		return nil, csvOptionsNotAllowed
+	}
+
+	tableWriteDisposition := bigquery.WriteAppend
+	if writeDisposition != nil {
+		tableWriteDisposition = *writeDisposition
+	}
+
+	return &Client{
+		client: client,
+
+		dataSetID: dataSetID,
+		tableID:   tableID,
+
+		schema:              schema,
+		sourceFormat:        sourceFormat,
+		autoDetect:          autoDetect,
+		ignoreUnknownValues: ignoreUnknownValues,
+		csvOptions:          csvOptions,
+		writeDisposition:    tableWriteDisposition,
+	}, nil
+}
+
+// Put implements bqClient::Put
+func (bqc *Client) Put(data interface{}) (bool, error) {
+	reader, ok := data.(io.Reader)
+	if !ok {
+		return false, couldNotConvertReader
+	}
+
+	ctx := context.Background()
+	source := bigquery.NewReaderSource(reader)
+	source.AutoDetect = bqc.autoDetect
+	source.SourceFormat = bqc.sourceFormat
+	source.IgnoreUnknownValues = bqc.ignoreUnknownValues
+	if bqc.csvOptions != nil {
+		source.CSVOptions = *bqc.csvOptions
+	}
+	if bqc.schema != nil {
+		source.Schema = *bqc.schema
+	}
+
+	table := bqc.client.Dataset(bqc.dataSetID).Table(bqc.tableID)
+	loader := table.LoaderFrom(source)
+	loader.WriteDisposition = bqc.writeDisposition
+	job, err := loader.Run(ctx)
+	if err != nil {
+		return false, fmt.Errorf("BQ batch client: failed to run loader %w", err)
+	}
+	status, err := job.Wait(ctx)
+	if err != nil {
+		return false, fmt.Errorf("BQ batch client: job failed while waiting %w", err)
+	}
+
+	if err := status.Err(); err != nil {
+		bqc.errors = status.Errors
+		return false, fmt.Errorf("BQ batch client: job returned an error status %w", err)
+	}
+
+	// we flush every time we write data.
+	return true, nil
+}
+
+// Flush implements bigquery.Client::Flush
+func (bqc *Client) Flush() error {
+	// NOTE: flushing is only equired once we support non-default streams
```

Comment:
this note is wrong, no? It should simply say that the data is always flushed upon putting data immediately, no? Or at least something like that

Reply:
I've updated the note, it was indeed a copy paste error that wasn't updated yet.